File: SOC_Photon/Battery State/EKF/sandbox/EKF_1x1.py
# BatteryEKF - general purpose battery class for embedded EKF
# Copyright (C) 2021 Dave Gutz
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation;
# version 2.1 of the License.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
# Lesser General Public License for more details.
#
# See http://www.fsf.org/licensing/licenses/lgpl.txt for full license text.

"""1x1 General Purpose Extended Kalman Filter.   Inherit from this class and include ekf_model_predict and
ekf_model_update methods in the parent."""


class EKF_1x1:
    """1x1 General Purpose Extended Kalman Filter.   Inherit from this class and include ekf_model_predict and
    ekf_model_update methods in the parent."""

    def __init__(self):
        """
        """
        self.Fx = 0.  # State transition
        self.Bu = 0.  # Control transition
        self.Q = 0.  # Process uncertainty
        self.R = 0.  # State uncertainty
        self.P = 0.  # Uncertainty covariance
        self.H = 0.  # Jacobian of h(x)
        self.H_eqn = 0.  # Jacobian of h(x)
        self.S = 0.  # System uncertainty
        self.K = 0.  # Kalman gain
        self.hx = 0.  # Output of observation function h(x)
        self.hx_eqn = 0.  # Output of observation function h(x)
        self.u_kf = 0.  # Control input
        self.x_kf = 0.  # Kalman state variable
        self.y_kf = 0.  # Residual z-hx
        self.z_ekf = 0.  # Observation of state x
        self.x_prior = self.x_kf
        self.P_prior = self.P
        self.x_post = self.x_kf
        self.P_post = self.P

    def __str__(self, prefix=''):
        """Returns representation of the object"""
        s = prefix + "EKF_1x1:\n"
        s += "  Inputs:\n"
        s += "  z = {:7.3f}\n".format(self.z_ekf)
        s += "  R = {:10.6f}\n".format(self.R)
        s += "  Q = {:10.6f}\n".format(self.Q)
        s += "  H = {:7.3f}\n".format(self.H)
        s += "  H_eqn = {:7.3f}\n".format(self.H_eqn)
        s += "  Outputs:\n"
        s += "  x  = {:7.3f}\n".format(self.x_kf)
        s += "  hx = {:7.3f}\n".format(self.hx)
        s += "  hx_eqn = {:7.3f}\n".format(self.hx_eqn)
        s += "  y  = {:7.3f}\n".format(self.y_kf)
        s += "  P  = {:10.6f}\n".format(self.P)
        s += "  K  = {:10.6f}\n".format(self.K)
        s += "  S  = {:10.6f}\n".format(self.S)
        return s

    def ekf_model_predict(self):
        raise NotImplementedError

    def ekf_model_update(self):
        raise NotImplementedError

    def init_ekf(self, soc, p_init):
        """Initialize on demand"""
        self.x_kf = soc
        self.P = p_init

    def predict_ekf(self, u):
        """1x1 Extended Kalman Filter predict
        Inputs:
            u   1x1 input, =ib, A
            Bu  1x1 control transition, Ohms
            Fx  1x1 state transition, V/V
        Outputs:
            x   1x1 Kalman state variable = Vsoc (0-1 fraction)
            P   1x1 Kalman probability
        """
        self.u_kf = u
        self.Fx, self.Bu = self.ekf_model_predict()
        self.x_kf = self.Fx*self.x_kf + self.Bu*self.u_kf
        self.P = self.Fx * self.P * self.Fx + self.Q
        self.x_prior = self.x_kf
        self.P_prior = self.P

    def update_ekf(self, z, x_min, x_max):
        """ 1x1 Extended Kalman Filter update
            Inputs:
                z   1x1 input, =voc, dynamic predicted by other model, V
                R   1x1 Kalman state uncertainty
                Q   1x1 Kalman process uncertainty
                H   1x1 Jacobian sensitivity dV/dSOC
            Outputs:
                x   1x1 Kalman state variable = Vsoc (0-1 fraction)
                y   1x1 Residual z-hx, V
                P   1x1 Kalman uncertainty covariance
                K   1x1 Kalman gain
                S   1x1 system uncertainty
                SI  1x1 system uncertainty inverse
        """
        self.hx, self.H = self.ekf_model_update()
        self.z_ekf = z
        pht = self.P*self.H
        self.S = self.H*pht + self.R
        if abs(self.S) > 1e-12:
            self.K = pht/self.S  # using last-good-value if S=0
        self.y_kf = self.z_ekf - self.hx
        self.x_kf = max(min(self.x_kf + self.K*self.y_kf, x_max), x_min)
        i_kh = 1 - self.K*self.H
        self.P = i_kh*self.P
        self.x_post = self.x_kf
        self.P_post = self.P

    def h_jacobian(self, x_kf):
        # implemented by child
        raise NotImplementedError

    def hx_calc(self):
        # implemented by child
        raise NotImplementedError

File: SOC_Particle/src/myLibrary/EKF_1x1.cpp
//
// MIT License
//
// Copyright (C) 2023 - Dave Gutz
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

#include "application.h"
#include "EKF_1x1.h"
#include <math.h>
#include "parameters.h"
extern SavedPars sp; // Various parameters to be static at system level and saved through power cycle
// extern int8_t debug();


// class EKF_1x1
// constructors
EKF_1x1::EKF_1x1(){}
EKF_1x1::~EKF_1x1() {}

// operators

// functions
//1x1 Extended Kalman Filter predict
void EKF_1x1::predict_ekf(double u)
{
  /*
  1x1 Extended Kalman Filter predict
  Inputs:
    u   1x1 input, =ib, A
    Bu  1x1 control transition, Ohms
    Fx  1x1 state transition, V/V
  Outputs:
    x   1x1 Kalman state variable = Vsoc (0-1 fraction)
    P   1x1 Kalman probability
  */
  u_ = u;
  this->ekf_predict(&Fx_, &Bu_);
  x_ = Fx_*x_ + Bu_*u_;
  if ( isnan(P_) ) P_ = 0.;   // reset overflow
  P_ = Fx_*P_*Fx_ + Q_;
  x_prior_ = x_;
  P_prior_ = P_;
}

// Initialize
void EKF_1x1::init_ekf(double soc, double Pinit)
{
  x_ = soc;
  P_ = Pinit;
}

// Pretty Print
 void EKF_1x1::pretty_print(void)
 {
#ifndef DEPLOY_PHOTON
  Serial.printf("EKF_1x1:\n");
  Serial.printf("In:\n");
  Serial.printf("  z  =   %8.4f, V\n", z_);
  Serial.printf("  R  = %10.6f\n", R_);
  Serial.printf("  Q  = %10.6f\n", Q_);
  Serial.printf("  H  =    %7.3f\n", H_);
  Serial.printf("Out:\n");
  Serial.printf("  x  =   %8.4f, Vsoc (0-1 fraction)\n", x_);
  Serial.printf("  hx =   %8.4f\n", hx_);
  Serial.printf("  y  =   %8.4f, V\n", y_);
  Serial.printf("  P  = %10.6f\n", P_);
  Serial.printf("  K  = %10.6f\n", K_);
  Serial.printf("  S  = %10.6f\n", S_);
#else
     Serial.printf("EKF_1x1: silent for DEPLOY_PHOTON\n");
#endif
 }

// Serial print
 void EKF_1x1::serial_print(const double control_time, const unsigned long int now, const float dt)
 {
  double cTime;
  if ( sp.tweak_test() ) cTime = double(now)/1000.;
  else cTime = control_time;
  Serial.printf("unit_ekf,%13.3f,%7.3f,%10.7g,%10.7g,%10.7g,%10.7g,%10.7g,%10.7g,%10.7g,%10.7g,%10.7g,%10.7g,%10.7g,%10.7g,%10.7g,%10.7g,%10.7g,%10.7g,%10.7g,\n",
    cTime, dt, Fx_, Bu_, Q_, R_, P_, S_, K_, u_, x_, y_, z_, x_prior_, P_prior_, x_post_, P_post_, hx_, H_);
 }

// y <- C@x + D@u
// Backward Euler integration of x
void EKF_1x1::update_ekf(const double z, double x_min, double x_max)
{
  /*1x1 Extended Kalman Filter update
  Inputs:
    z   1x1 input, =voc, dynamic predicted by other model, V
    R   1x1 Kalman state uncertainty
    Q   1x1 Kalman process uncertainty
    H   1x1 Jacobian sensitivity dV/dSOC
  Outputs:
    x   1x1 Kalman state variable = Vsoc (0-1 fraction)
    hx  1x1 Output of observation function h(x)
    y   1x1 Residual z-hx, V
    P   1x1 Kalman uncertainty covariance
    K   1x1 Kalman gain
    S   1x1 system uncertainty
    SI  1x1 system uncertainty inverse
  */
  this->ekf_update(&hx_, &H_);
  z_ = z;
  double pht = P_*H_;
  S_ = H_*pht + R_;
  if ( abs(S_) > 1e-12) K_ = pht / S_;  // Using last-good-value if S_ = 0
  y_ = z_ - hx_;
  x_ = max(min( x_ + K_*y_, x_max), x_min);
  double i_kh = 1. - K_*H_;
  P_ *= i_kh;
  x_post_ = x_;
  P_post_ = P_;
}
